keps/sig-storage/20191117-noupdatable-volumes.md

```diff
@@ -0,0 +1,177 @@
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+The most popular and the most convinient way of consuming Secrets and
+ConfigMaps by Pods is consuming it as a file. However, any update to
+a Secret or ConfigMap object is quickly (roughly within a minute)
+reflected in updates of the file mounted for *all* Pods consuming them.
+That means, that a bad update (push) of Secret and/or ConfigMap can
+very quickly break the whole appplication.
+
+The recommended approach for upgrading applications are obviously
+rolling upgrades. For Secrets/ConfigMaps this is translating to
+creating a new object and updating PodTemplate with the reference
+to it. However, that doesn't protect users from outages caused by
+accidental bad updates of existing Secrets and/or ConfigMaps.
+
+Moreover, the feature of updating Secrets/ConfigMaps for running
+Pods is also expensive from scalability/performance point of view.
+Every Kubelet has to watch (default) or periodically poll every
+unique Secret and ConfigMap referenced by any of the Pods it is
+running.
+
+## Motivation
+
+In this KEP, we are proposing introducing an ability to specify that
+particular Secrets/ConfigMaps should not be updated at runtime for a
+given Pods.
+Given there are a lot of users not really taking advantage of automatic
+updates of Secrets/ConfigMaps due to consequences described above, this
+will allow then to:
+- protect themselves better for accidental bad updates that could cause
+  outages of their applications
+- achieve better performance of their cluster thanks to significant
+  reduction of load on apiserver
+
+### Goals
+
+- introduce protection mechanism to avoid outages due to accidental
+  updates of existing Secrets/ConfigMaps
+- improve cluster performance by reducing load on Kubernetes control
+  plance (mostly kube-apiserver) consumed by not-utilized feature
```

Comment:
With watch based secrets, do we actually have apiserver load issues?

I've actually seen that when you have a few hundred secrets, the node itself incurs significant cpu just maintaining steady state to diff between old and new secrets. There probably could be optimizations there by caching last updated time so that we only diff on a change

Reply:
> With watch based secrets, do we actually have apiserver load issues?

We do (different ones, but still).
The issue is that with 5k nodes and 150k pods, we are getting to O(hundreds of thousands) open watches. And that is still causing some issues on apiserver (mostly connected to memory allocations at the go level).
So that would actually visibly improve our situation too.